File: 3373.maximize-the-number-of-target-nodes-after-connecting-trees-ii.java
#
# @lc app=leetcode id=3373 lang=java
#
# [3373] Maximize the Number of Target Nodes After Connecting Trees II
#

# @lc code=start
import java.util.*;

class Solution {
    public int[] maxTargetNodes(int[][] edges1, int[][] edges2) {
        int n = edges1.length + 1;
        int m = edges2.length + 1;

        // Build adj1 and compute colors and counts for tree1
        List<Integer>[] adj1 = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj1[i] = new ArrayList<>();
        }
        for (int[] e : edges1) {
            int a = e[0], b = e[1];
            adj1[a].add(b);
            adj1[b].add(a);
        }
        int[] color1 = new int[n];
        boolean[] vis1 = new boolean[n];
        Queue<Integer> q1 = new LinkedList<>();
        q1.offer(0);
        vis1[0] = true;
        color1[0] = 0;
        int cnt0_1 = 1;
        int cnt1_1 = 0;
        while (!q1.isEmpty()) {
            int u = q1.poll();
            for (int v : adj1[u]) {
                if (!vis1[v]) {
                    vis1[v] = true;
                    color1[v] = 1 - color1[u];
                    if (color1[v] == 0) cnt0_1++;
                    else cnt1_1++;
                    q1.offer(v);
                }
            }
        }

        // Build adj2 and compute counts for tree2
        List<Integer>[] adj2 = new ArrayList[m];
        for (int i = 0; i < m; i++) {
            adj2[i] = new ArrayList<>();
        }
        for (int[] e : edges2) {
            int a = e[0], b = e[1];
            adj2[a].add(b);
            adj2[b].add(a);
        }
        int[] color2 = new int[m];
        boolean[] vis2 = new boolean[m];
        Queue<Integer> q2 = new LinkedList<>();
        q2.offer(0);
        vis2[0] = true;
        color2[0] = 0;
        int cnt0_2 = 1;
        int cnt1_2 = 0;
        while (!q2.isEmpty()) {
            int u = q2.poll();
            for (int v : adj2[u]) {
                if (!vis2[v]) {
                    vis2[v] = true;
                    color2[v] = 1 - color2[u];
                    if (color2[v] == 0) cnt0_2++;
                    else cnt1_2++;
                    q2.offer(v);
                }
            }
        }
        int max2 = Math.max(cnt0_2, cnt1_2);

        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            ans[i] = (color1[i] == 0 ? cnt0_1 : cnt1_1) + max2;
        }
        return ans;
    }
}
# @lc code=end

File: 3373.maximize-the-number-of-target-nodes-after-connecting-trees-ii.go
#
# @lc app=leetcode id=3373 lang=golang
#
# [3373] Maximize the Number of Target Nodes After Connecting Trees II
#

# @lc code=start
func maxTargetNodes(edges1 [][]int, edges2 [][]int) []int {

	bfs := func(adj [][]int, nn int) []int {
		color := make([]int, nn)
		for i := range color {
			color[i] = -1
		}
		q := []int{}
		qi := 0
		color[0] = 0
		q = append(q, 0)
		for qi < len(q) {
			u := q[qi]
			qi++
			for _, v := range adj[u] {
				if color[v] == -1 {
					color[v] = 1 - color[u]
					q = append(q, v)
				}
			}
		}
		return color
	}

	n := len(edges1) + 1
	m := len(edges2) + 1

	adj1 := make([][]int, n)
	for _, e := range edges1 {
		u, v := e[0], e[1]
		adj1[u] = append(adj1[u], v)
		adj1[v] = append(adj1[v], u)
	}

	adj2 := make([][]int, m)
	for _, e := range edges2 {
		u, v := e[0], e[1]
		adj2[u] = append(adj2[u], v)
		adj2[v] = append(adj2[v], u)
	}

	color1 := bfs(adj1, n)
	cnt0_1, cnt1_1 := 0, 0
	for _, c := range color1 {
		if c == 0 {
			cnt0_1++
		} else {
			cnt1_1++
		}
	}

	color2 := bfs(adj2, m)
	cnt0_2, cnt1_2 := 0, 0
	for _, c := range color2 {
		if c == 0 {
			cnt0_2++
		} else {
			cnt1_2++
		}
	}

	max2 := cnt0_2
	if cnt1_2 > max2 {
		max2 = cnt1_2
	}

	ans := make([]int, n)
	for i := 0; i < n; i++ {
		if color1[i] == 0 {
			ans[i] = cnt0_1 + max2
		} else {
			ans[i] = cnt1_1 + max2
		}
	}

	return ans
}
# @lc code=end